### ISSUE TITLE
Bug: [no-unnecessary-condition] don't flag values of an unconstrained or valid type parameter

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10442]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3hkmCMAXPABQCU8AvAHzwBGSSEIUKA3AFCiSwLLpY5MAJiJlKNAK4oA1iiQB3Np3DQ4iVJmy4MAZiIAeACpUx1eIY5dVvDQO0AWA4fggAHjhTA0dBkxYnyMwt2dgB6UPgAByg0NFx2AG08ABp4FAkICABdADoAMwBLCBwYYkEMPFIOMIjo2Pik1PTM3MLikFLyoSqQ8Pg0AAskDOAomLRElLSM7PyikrLtHR6a-qGRsdjJppnW%2BY7FzHseoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const test1: () => boolean;
declare const test2: () => unknown;
declare const test3: <T>() => T;
declare const test4: <T extends boolean>() => T;

// passes
[1, null].filter(test1);

// passes
[1, null].filter(test2);

// should pass
[1, null].filter(test3);

// should pass
[1, null].filter(test4);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect all use-cases to be consistent and not report as errors.

### Actual Result

The use-cases with a type constraint report (unnecessarily) as an error.

### Additional Info

This seems similar to #10311 and looks like the fix should be similar.
